src/app/ui/palette_listbox.cpp

```diff
@@ -135,90 +141,100 @@ class PaletteListBox::LoadingItem : public ListItem {
+  auto resource = selectedPaletteResource();
+  return resource ? resource->palette() : nullptr;
+}
 
-  }
-  return ListBox::onProcessMessage(msg);
+std::string PaletteListBox::selectedPaletteName() {
+  auto resource = selectedPaletteResource();
+  return resource ? resource->name() : "";
 }
 
-void PaletteListBox::onChange()
-{
-  doc::Palette* palette = selectedPalette();
+void PaletteListBox::onChange() {
+  auto palette = selectedPalette();
   if (palette)
     PalChange(palette);
 }
 
-void PaletteListBox::onTick()
-{
-  if (m_resourcesLoader == NULL) {
-    stop();
-    return;
-  }
-
-  if (!m_loadingItem) {
-    m_loadingItem = new LoadingItem;
-    addChild(m_loadingItem);
-  }
-  m_loadingItem->makeProgress();
-
-  base::UniquePtr<Resource> resource;
-  std::string name;
-
-  if (!m_resourcesLoader->next(resource)) {
-    if (m_resourcesLoader->isDone()) {
-      stop();
-
-      LOG("Done\n");
+void PaletteListBox::setLoading(bool isLoading) {
+  if (isLoading) {
+    if (!m_loadingItem) {
+      m_loadingItem = new LoadingItem;
+      addChild(m_loadingItem);
+    }
+    m_loadingItem->makeProgress();
+  } else {
+    if (m_loadingItem) {
+      removeChild(m_loadingItem);
+      delete m_loadingItem;
+      m_loadingItem = nullptr;
+      invalidate();
     }
-    return;
   }
+}
 
-  PaletteResource* palResource = static_cast<PaletteResource*>(resource.get());
-  base::UniquePtr<PaletteListItem> listItem(new PaletteListItem(palResource));
-  insertChild(getItemsCount()-1, listItem);
+void PaletteListBox::addPalette(doc::Palette *palette, const std::string& name) {
+  int hasLoading = !!m_loadingItem;
+  insertChild(getItemsCount()-hasLoading, new PaletteListItem(std::unique_ptr<doc::Palette>(palette), name));
   layout();
+  View* view = View::getView(this);
+  if (view)
+    view->updateView();
+}
 
+void PaletteListBox::addPalette(PaletteResource *resource) {
+  int hasLoading = !!m_loadingItem;
+  insertChild(getItemsCount()-hasLoading, new PaletteListItem(resource));
+  layout();
   View* view = View::getView(this);
   if (view)
     view->updateView();
+}
 
-  resource.release();
-  listItem.release();
+PaletteFileListBox::PaletteFileListBox() : m_resourcesLoader(new ResourcesLoader(new PalettesLoaderDelegate)),
+                                           m_resourcesTimer(100) {
+  m_resourcesTimer.Tick.connect(base::Bind<void>(&PaletteFileListBox::onTick, this));
 }
 
-void PaletteListBox::stop()
-{
-  if (m_loadingItem) {
-    removeChild(m_loadingItem);
-    delete m_loadingItem;
-    m_loadingItem = NULL;
+bool PaletteFileListBox::onProcessMessage(ui::Message* msg) {
+  switch (msg->type()) {
+    case kOpenMessage: {
+      m_resourcesTimer.start();
+      break;
+    }
+  }
+  return ListBox::onProcessMessage(msg);
+}
 
-    invalidate();
+void PaletteFileListBox::onTick() {
+  if (m_resourcesLoader == NULL) {
+    stop();
+    return;
   }
+  setLoading(true);
+  while (true) {
+    base::UniquePtr<Resource> resource;
```

Comment:
any reason why this (and a few more) are still `base::UniquePtr` instead of `std::unique_ptr`?

Reply:
ResourcesLoader requires a base::UniquePtr so I couldn't change without breaking a lot of other things.